File: packages/quran-engine-swift/Sources/QuranEngine/Morphology.swift
import Foundation

/// Root and lemma of every word of the Quran: the Quranic Arabic Corpus morphology (Kais Dukes),
/// as redistributed by the Quranic Universal Library.
///
/// The invariant, the same one `word-by-word.json` keeps: one id per whitespace token of the
/// ayah's raw Hafs text, in the text's own token order, so nothing here matches or normalizes
/// text. Id `0` means the token has neither a root nor a lemma, which is the honest answer for
/// particles and the sajdah mark. Ids are 1-based into the root and lemma tables.
///
/// The reverse indexes are built on first use and kept: walking 77,629 tokens is quick, but a
/// caller asking about ten roots should not pay for it ten times.
///
/// See docs/18-morphology.md.

/// A triliteral (or quadriliteral) root, in both spellings a reader might use.
public struct MorphologyRoot: Sendable, Equatable {
    /// Spaced the way a lexicon prints it: "ر ب ب".
    public let letters: String
    /// The same letters transliterated: "rbb".
    public let buckwalter: String
    /// `letters` with the spaces closed up: "ربب".
    public var joined: String { letters.replacingOccurrences(of: " ", with: "") }
}

/// A dictionary form, marked and unmarked.
public struct MorphologyLemma: Sendable, Equatable {
    public let text: String
    public let clean: String
}

/// One word of the Quran: the ayah, and the 0-based index of the token inside it.
public struct WordLocation: Sendable, Equatable, Hashable {
    public let surah: Int
    public let ayah: Int
    public let token: Int

    public init(surah: Int, ayah: Int, token: Int) {
        self.surah = surah
        self.ayah = ayah
        self.token = token
    }
}

/// `data/morphology.json`.
public struct MorphologyFile: Decodable, Sendable {
    public let roots: [[String]]
    public let lemmas: [[String]]
    public let rootIds: [String: [[Int]]]
    public let lemmaIds: [String: [[Int]]]
}

public final class Morphology: @unchecked Sendable {
    private let file: MorphologyFile?
    private let lock = NSLock()
    private var byRoot: [Int: [WordLocation]]?
    private var byLemma: [Int: [WordLocation]]?

    public init(_ file: MorphologyFile? = nil) {
        self.file = file
    }

    /// Whether a corpus was supplied at all.
    public var isLoaded: Bool { !(file?.roots.isEmpty ?? true) }

    /// The fold a typed query and the tables are both compared under.
    ///
    /// Every space is removed, not merely trimmed: a root is printed spaced ("ر ب ب") and typed
    /// closed up ("ربب"), and the two have to meet.
    public static func fold(_ text: String) -> String {
        ArabicText.cleanSearch(ArabicText.removingArabicDiacriticsAndSigns(text))
            .components(separatedBy: .whitespacesAndNewlines)
            .joined()
    }

    /// The root with this 1-based id.
    public func root(id: Int) -> MorphologyRoot? {
        guard let file, id >= 1, id <= file.roots.count else { return nil }
        let row = file.roots[id - 1]
        guard row.count >= 2 else { return nil }
        return MorphologyRoot(letters: row[0], buckwalter: row[1])
    }

    /// The dictionary form with this 1-based id.
    public func lemma(id: Int) -> MorphologyLemma? {
        guard let file, id >= 1, id <= file.lemmas.count else { return nil }
        let row = file.lemmas[id - 1]
        guard row.count >= 2 else { return nil }
        return MorphologyLemma(text: row[0], clean: row[1])
    }

    /// The root and lemma id of every token of the ayah, or nil when it is not covered.
    public func ids(surah: Int, ayah: Int) -> (roots: [Int], lemmas: [Int])? {
        guard let file,
              let roots = file.rootIds["\(surah)"], let lemmas = file.lemmaIds["\(surah)"],
              ayah >= 1, ayah <= roots.count, ayah <= lemmas.count else { return nil }
        return (roots[ayah - 1], lemmas[ayah - 1])
    }

    /// The root of one token; nil when it has none (a particle) or the index is out of range.
    public func root(surah: Int, ayah: Int, token: Int) -> (id: Int, root: MorphologyRoot)? {
        guard let ids = ids(surah: surah, ayah: ayah), ids.roots.indices.contains(token),
              let found = root(id: ids.roots[token]) else { return nil }
        return (ids.roots[token], found)
    }

    /// The dictionary form of one token.
    public func lemma(surah: Int, ayah: Int, token: Int) -> (id: Int, lemma: MorphologyLemma)? {
        guard let ids = ids(surah: surah, ayah: ayah), ids.lemmas.indices.contains(token),
              let found = lemma(id: ids.lemmas[token]) else { return nil }
        return (ids.lemmas[token], found)
    }

    /// Every word carrying this root, in mushaf order.
    public func occurrences(ofRoot id: Int) -> [WordLocation] {
        buildIndex()
        lock.lock(); defer { lock.unlock() }
        return byRoot?[id] ?? []
    }

    /// Every word carrying this lemma, in mushaf order.
    public func occurrences(ofLemma id: Int) -> [WordLocation] {
        buildIndex()
        lock.lock(); defer { lock.unlock() }
        return byLemma?[id] ?? []
    }

    /// Roots whose Arabic or Buckwalter spelling starts with the query.
    public func findRoots(_ query: String, limit: Int = 50) -> [(id: Int, root: MorphologyRoot)] {
        prefixHits(file?.roots ?? [], query, limit).compactMap { id in
            root(id: id).map { (id, $0) }
        }
    }

    /// Dictionary forms whose marked or unmarked spelling starts with the query.
    public func findLemmas(_ query: String, limit: Int = 50) -> [(id: Int, lemma: MorphologyLemma)] {
        prefixHits(file?.lemmas ?? [], query, limit).compactMap { id in
            lemma(id: id).map { (id, $0) }
        }
    }

    /// Corpus size: roots, lemmas, and the tokens they cover.
    public func count() -> (roots: Int, lemmas: Int, tokens: Int) {
        guard let file else { return (0, 0, 0) }
        let tokens = file.rootIds.values.reduce(0) { total, ayahs in
            total + ayahs.reduce(0) { $0 + $1.count }
        }
        return (file.roots.count, file.lemmas.count, tokens)
    }

    // MARK: - Internals

    private func prefixHits(_ table: [[String]], _ query: String, _ limit: Int) -> [Int] {
        let folded = Self.fold(query)
        let latin = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !folded.isEmpty || !latin.isEmpty else { return [] }
        var out: [Int] = []
        for (index, row) in table.enumerated() {
            if limit > 0 && out.count >= limit { break }
            guard let first = row.first else { continue }
            let arabic = Self.fold(first)
            let roman = (row.count > 1 ? row[1] : "").lowercased()
            if (!folded.isEmpty && arabic.hasPrefix(folded))
                || (!latin.isEmpty && roman.hasPrefix(latin)) {
                out.append(index + 1)
            }
        }
        return out
    }

    /// Walk the corpus once, in mushaf order, so the lists come out ordered for free.
    private func buildIndex() {
        lock.lock(); defer { lock.unlock() }
        guard byRoot == nil, let file else { return }
        var roots: [Int: [WordLocation]] = [:]
        var lemmas: [Int: [WordLocation]] = [:]
        for surah in file.rootIds.keys.compactMap(Int.init).sorted() {
            let rootRows = file.rootIds["\(surah)"] ?? []
            let lemmaRows = file.lemmaIds["\(surah)"] ?? []
            for (index, rootRow) in rootRows.enumerated() {
                let lemmaRow = index < lemmaRows.count ? lemmaRows[index] : []
                for (token, rootID) in rootRow.enumerated() {
                    let location = WordLocation(surah: surah, ayah: index + 1, token: token)
                    if rootID != 0 { roots[rootID, default: []].append(location) }
                    if token < lemmaRow.count, lemmaRow[token] != 0 {
                        lemmas[lemmaRow[token], default: []].append(location)
                    }
                }
            }
        }
        byRoot = roots
        byLemma = lemmas
    }
}

File: packages/quran-engine-js/src/morphology.js
// @ts-check
/**
 * Root and lemma of every word of the Quran: the Quranic Arabic Corpus morphology (Kais Dukes),
 * as redistributed by the Quranic Universal Library.
 *
 * Two questions this answers that nothing else in the engine can. Given a word, what is its root
 * and its dictionary form, and where else does that root appear. Given a typed root (رحم, or
 * `rHm` in Buckwalter), which ayahs carry it. That second one is why a search box can answer a
 * bare root at all: matching text would find only the surface forms that happen to be spelled
 * that way.
 *
 * **The invariant, the same one `word-by-word.json` keeps**: one id per whitespace token of the
 * ayah's raw Hafs text, in the text's own token order. The alignment against Quran.com's word
 * positions happened at build time, so nothing here matches or normalizes text. Index `n` is the
 * `n`th token; id `0` means the token has neither a root nor a lemma, which is the honest answer
 * for particles and for the sajdah mark. Ids are 1-based into `roots` and `lemmas`.
 *
 * The occurrence indexes are built on first use and then kept: walking 77,629 tokens is fast, but
 * a caller asking for every occurrence of ten different roots should not pay for it ten times.
 */

import { cleanSearch, removingArabicDiacriticsAndSigns } from "./text.js";

/**
 * @typedef {Object} Root
 * @property {string} letters      the letters spaced the way a lexicon prints them, "ر ب ب"
 * @property {string} buckwalter   the same letters in Buckwalter transliteration, "rbb"
 * @property {string} joined       the letters with the spaces closed up, "ربب"
 */

/**
 * @typedef {Object} Lemma
 * @property {string} text   the dictionary form with its marks, "رَبّ"
 * @property {string} clean  the same form unmarked, "رب"
 */

/**
 * @typedef {Object} WordLocation
 * @property {number} surah
 * @property {number} ayah
 * @property {number} token  0-based index into the ayah's whitespace tokens
 */

/**
 * The fold a typed query and the tables are both compared under: marks and spaces gone.
 *
 * Every space is removed, not merely trimmed, because a root is printed spaced ("ر ب ب") and
 * typed closed up ("ربب"), and the two have to meet. Note that `cleanSearch`'s own `whitespace`
 * option only trims the ends, so the closing-up is done here.
 */
export function foldForMorphology(text) {
  return cleanSearch(removingArabicDiacriticsAndSigns(text ?? ""), { whitespace: true })
    .replace(/\s+/g, "");
}

export class Morphology {
  /**
   * @param {Object} [data] data/morphology.json
   * @param {Array<[string, string]>} [data.roots]
   * @param {Array<[string, string]>} [data.lemmas]
   * @param {Record<string, number[][]>} [data.rootIds]
   * @param {Record<string, number[][]>} [data.lemmaIds]
   */
  constructor(data = {}) {
    this._roots = data.roots ?? [];
    this._lemmas = data.lemmas ?? [];
    this._rootIds = data.rootIds ?? {};
    this._lemmaIds = data.lemmaIds ?? {};
    /** @type {Map<number, WordLocation[]>|null} */
    this._byRoot = null;
    /** @type {Map<number, WordLocation[]>|null} */
    this._byLemma = null;
  }

  /** Whether a corpus was supplied at all. */
  get isLoaded() {
    return this._roots.length > 0;
  }

  /** @param {number} id @returns {Root|null} */
  root(id) {
    const row = id >= 1 ? this._roots[id - 1] : undefined;
    if (!row) return null;
    return { letters: row[0], buckwalter: row[1], joined: row[0].replace(/ /g, "") };
  }

  /** @param {number} id @returns {Lemma|null} */
  lemma(id) {
    const row = id >= 1 ? this._lemmas[id - 1] : undefined;
    return row ? { text: row[0], clean: row[1] } : null;
  }

  /**
   * The root and lemma id of every token of the ayah, or null when the ayah is not covered.
   * @param {number} surahId @param {number} ayahId
   * @returns {{roots: number[], lemmas: number[]}|null}
   */
  ids(surahId, ayahId) {
    const roots = this._rootIds[String(surahId)];
    const lemmas = this._lemmaIds[String(surahId)];
    if (!roots || !lemmas || ayahId < 1 || ayahId > roots.length || ayahId > lemmas.length) return null;
    return { roots: roots[ayahId - 1], lemmas: lemmas[ayahId - 1] };
  }

  /**
   * The root of one token, or null when the token has none (a particle) or is out of range.
   * @param {number} surahId @param {number} ayahId @param {number} token
   * @returns {{id: number, root: Root}|null}
   */
  rootOf(surahId, ayahId, token) {
    const ids = this.ids(surahId, ayahId);
    if (!ids || token < 0 || token >= ids.roots.length) return null;
    const id = ids.roots[token];
    const root = this.root(id);
    return root ? { id, root } : null;
  }

  /**
   * The dictionary form of one token.
   * @param {number} surahId @param {number} ayahId @param {number} token
   * @returns {{id: number, lemma: Lemma}|null}
   */
  lemmaOf(surahId, ayahId, token) {
    const ids = this.ids(surahId, ayahId);
    if (!ids || token < 0 || token >= ids.lemmas.length) return null;
    const id = ids.lemmas[token];
    const lemma = this.lemma(id);
    return lemma ? { id, lemma } : null;
  }

  /** Every word carrying this root, in mushaf order. @param {number} id @returns {WordLocation[]} */
  occurrencesOfRoot(id) {
    this._index();
    return this._byRoot?.get(id) ?? [];
  }

  /** Every word carrying this lemma, in mushaf order. @param {number} id @returns {WordLocation[]} */
  occurrencesOfLemma(id) {
    this._index();
    return this._byLemma?.get(id) ?? [];
  }

  /**
   * Roots matching a typed query, by prefix of either spelling. Arabic is folded, so a query
   * with or without marks and with or without spaces finds the same root.
   * @param {string} query @param {number} [limit]
   * @returns {Array<{id: number, root: Root}>}
   */
  findRoots(query, limit = 50) {
    return this._find(this._roots, query, limit, (row) => ({
      letters: row[0], buckwalter: row[1], joined: row[0].replace(/ /g, ""),
    }));
  }

  /**
   * Lemmas matching a typed query, by prefix of the marked or unmarked form.
   * @param {string} query @param {number} [limit]
   * @returns {Array<{id: number, lemma: Lemma}>}
   */
  findLemmas(query, limit = 50) {
    const hits = this._find(this._lemmas, query, limit, (row) => ({ text: row[0], clean: row[1] }));
    return hits.map(({ id, root }) => ({ id, lemma: /** @type any */ (root) }));
  }

  /** Corpus size: roots, lemmas, and the tokens they cover. */
  count() {
    let tokens = 0;
    for (const ayahs of Object.values(this._rootIds)) for (const row of ayahs) tokens += row.length;
    return { roots: this._roots.length, lemmas: this._lemmas.length, tokens };
  }

  /** @private */
  _find(table, query, limit, shape) {
    const folded = foldForMorphology(query);
    const latin = (query ?? "").trim().toLowerCase();
    if (!folded && !latin) return [];
    const out = [];
    for (let i = 0; i < table.length && out.length < limit; i++) {
      const row = table[i];
      const arabic = foldForMorphology(row[0]);
      const roman = String(row[1] ?? "").toLowerCase();
      if ((folded && arabic.startsWith(folded)) || (latin && roman.startsWith(latin))) {
        out.push({ id: i + 1, root: shape(row) });
      }
    }
    return out;
  }

  /** @private Build the reverse indexes once, in mushaf order. */
  _index() {
    if (this._byRoot) return;
    /** @type {Map<number, WordLocation[]>} */
    const byRoot = new Map();
    /** @type {Map<number, WordLocation[]>} */
    const byLemma = new Map();
    const surahs = Object.keys(this._rootIds).map(Number).sort((a, b) => a - b);
    for (const surah of surahs) {
      const roots = this._rootIds[String(surah)] ?? [];
      const lemmas = this._lemmaIds[String(surah)] ?? [];
      for (let a = 0; a < roots.length; a++) {
        const rootRow = roots[a];
        const lemmaRow = lemmas[a] ?? [];
        for (let t = 0; t < rootRow.length; t++) {
          const location = { surah, ayah: a + 1, token: t };
          const rootId = rootRow[t];
          if (rootId) {
            const bucket = byRoot.get(rootId);
            if (bucket) bucket.push(location); else byRoot.set(rootId, [location]);
          }
          const lemmaId = lemmaRow[t];
          if (lemmaId) {
            const bucket = byLemma.get(lemmaId);
            if (bucket) bucket.push(location); else byLemma.set(lemmaId, [location]);
          }
        }
      }
    }
    this._byRoot = byRoot;
    this._byLemma = byLemma;
  }
}
